src/audio/tdfb/tdfb_ipc4.c

```diff
@@ -54,32 +49,38 @@ static struct ipc_msg *tdfb_notification_init(struct processing_module *mod,
 	primary->r.rsp = SOF_IPC4_MESSAGE_DIR_MSG_REQUEST;
 	primary->r.msg_tgt = SOF_IPC4_MESSAGE_TARGET_FW_GEN_MSG;
 	msg = ipc_msg_w_ext_init(msg_proto.header, msg_proto.extension,
-				 sizeof(struct tdfb_notification_payload));
+				 sizeof(struct sof_ipc4_notify_module_data) +
+				 sizeof(struct sof_ipc4_control_msg_payload) +
+				 sizeof(struct sof_ipc4_ctrl_value_chan));
 	if (!msg)
 		return NULL;
 
-	payload = (struct tdfb_notification_payload *)msg->tx_data;
-	payload->module_data.instance_id = IPC4_INST_ID(ipc_config->id);
-	payload->module_data.module_id = IPC4_MOD_ID(ipc_config->id);
-	payload->module_data.event_id = SOF_IPC4_NOTIFY_MODULE_EVENTID_ALSA_MAGIC_VAL |
+	msg_module_data = (struct sof_ipc4_notify_module_data *)msg->tx_data;
+	msg_module_data->instance_id = IPC4_INST_ID(ipc_config->id);
+	msg_module_data->module_id = IPC4_MOD_ID(ipc_config->id);
+	msg_module_data->event_id = SOF_IPC4_NOTIFY_MODULE_EVENTID_ALSA_MAGIC_VAL |
 		control_type_param_id;
-	payload->module_data.event_data_size = sizeof(struct sof_ipc4_control_msg_payload) +
+	msg_module_data->event_data_size = sizeof(struct sof_ipc4_control_msg_payload) +
 		sizeof(struct sof_ipc4_ctrl_value_chan);
-	payload->control_msg.id = control_id;
-	payload->control_msg.num_elems = 1;
-	payload->control_value.channel = 0;
+
+	msg_payload = (struct sof_ipc4_control_msg_payload *)msg_module_data->event_data;
```

Comment:
@singalsu I think this needs to be on the heap and copied as the lifetime of IPC message data is finite until the next message ?

Reply:
The message is allocated with rzalloc() by previous call msg = ipc_msg_w_ext_init(). see https://github.com/thesofproject/sof/blob/64cb13364e778f11b8404443c9bb233a82bc42b9/src/include/sof/ipc/msg.h#L53 .

Then msg_module_data is set to msg->tx_data that is rzallocated too in the same function for parameter size.

msg_payload is then set to msg_module_data->event_data. And ipc_msg_free() is called from tdfb_free().

I can't see from above what I'm missing. Please help!